Automattic/wp-calypso
# Add site channels for realtime

## Changes proposed in this Pull Request
The goal is to have real time comments appear in all Reader sections while keeping the number of joined channels per socket to a minimum. We are using one lasagna channel per site instead of having a channel for each post, that should help us reduce the number of joined channels and join/leave actions.

### Concepts
Lasagna is built using [Phoenix](https://www.phoenixframework.org/) which gives us multiplexing capabilities allowing one socket connection to handle multiple concerns. At the moment we are having two types of channels (concerns) that are handled via the same websocket. 


#### Sockets
At the moment we are connecting to the Lasagna socket only inside Reader, which means whenever a user navigates to the Reader (or one of its subsections) a socket is opened. Leaving the reader will close that socket. This happens in the [lasagna middleware](https://github.com/Automattic/wp-calypso/blob/789b060fa4b15f4d70a969985db1b28f8e5c0be4/client/state/lasagna/middleware.js) and it will make sure we only have one socket open at a time.

**Lifecycle**
We keep track of the socket connection status with the following states

- `init` socket was code was initiliazed
- `opening` socket is in the process of connecting
- `open_error` there was an error during connection step
- `opened` socket is opened for communication
- `closed` socket is closed for communication


#### Channels
There are currently two different [channel](https://hexdocs.pm/phoenix/channels.html) types (topics) that are supported for realtime communication.

**User Channels**

User specific channel using a topic like `user:wpcom:<userId>` it will be joined when the socket is connected and will be left on socket disconnect. At this point this is channel is not used but it could be easily handling any realtime user specific concerns.

**Site Channels**

Site specific channel using a topic like `public:push:site:<siteId>`, it is used to transmit per site updates. A user can join multiple site channels depending on how many sites are of interested. Currently these are used only in the Reader section, to update comments in realtime. 

These are rules for when a user can join or leave a site channel:

- A site channel can be joined once
- A site channel will be kept active for one hour after the join, to make sure we avoid unnecessary join/leave as the user is scrolling through articles
- A site channel can be left after 45 minutes of inactivity (no new messages)
- We can have a maximum of 20 site channels open at the same time, if a new channel has to be joined, the oldest (first joined) will be removed (left)
- Site channels that have posts that are currently viewed by the user will be kept active even if they don't meet the above requirements

We have a special [management](https://github.com/Automattic/wp-calypso/blob/68d7ac0778b42519e6258bc49d2d704afef0b68f/client/state/lasagna/site-channel/manager.js) for these channels to make sure we don't have too many "opened" (joined) at the same time.

#### Feed full post view
Whenever the user clicks on a post from [Following Sites](http://calypso.localhost:3000/read) it will be sent to a full post view for that feed item (url like `/read/feeds/<feedId>/posts/<itemId>`) at this point if the socket it is not connected we will attempt that and then join the site channel of this particular post.

#### Site full post view
Whenever the user clicks on a post from [Conversations](http://calypso.localhost:3000/read/conversations) it will be sent to a full post view for that blog item (url like `/read/blog/<blogId>/posts/<postId>`) at this point if the socket it is not connected we will attempt that and then join the blog channel of this particular post.

#### Conversation
For the [Conversations](http://calypso.localhost:3000/read/conversations) sections we will connect to the socket as soon as possible and join the channels for the visible site posts, while scrolling we will detect visible site posts and join their channel. We keep the currently viewed site and post ids in `redux/reader/viewing`.


## To do
- We currently don't update the comment counts in non conversational feeds like Following Sites](http://calypso.localhost:3000/read), to be fixed in a future PR
- We don't have any visual cues to help the user notice a new comment was added, to be fixed in a future PR
- Check backend new comment filters to make sure we are sending the new comments after they pass the spam filter
- Polling for new posts would show a `New Posts` sticker which on click would populate the feed with a new duplicated post containing new comments
- Handle comment updates and deletes in realtime

## Testing instructions
- Apply patch D40881-code
- Update line 81 in `wp-content/mu-plugins/lasagna.php` to
```
'ns'         => 'site' . ':' . $blog_id,
```
- Sandbox one of your sites, follow it and start a conversation (multiple comments and replies)
- `npm install`
- Go to http://calypso.localhost:3000/read/conversations
- After you see the conversation go to your site, add another comment
- It should appear in the reader


### Edge cases
In development we expose `window.SOCKET` and `window.CHANNELS` to help us see what is the current state of our socket and joined channels.

We also have lots of debug logs that can be followed to make sure channel management works correctly, you can enable those by running `localStorage.debug = "lasagna:*"` in the browser console.

**Site channel conditions for join/leave should be tested carefully, we could update the [constants](https://github.com/Automattic/wp-calypso/blob/68d7ac0778b42519e6258bc49d2d704afef0b68f/client/state/lasagna/site-channel/manager.js#L17) for easier testability, decreasing the number of max channel or the lifetime of a channel.**

**Feed Full Post**

- Hard refreshing a feed full post https://d.pr/v/sloTzj
- Navigating to feed full post from the feed list https://d.pr/v/Ic33MK
- Related redux state https://d.pr/v/Nddnc4

**Feed Site Post**

- Hard refreshing the site full post + leave https://d.pr/v/yjoNoa
- Navigating to site full post from Conversations https://d.pr/v/Dl73J8
- Related redux state https://d.pr/v/YnoZXd

**Conversations**

When first joining this there is a blip when more posts are visible until all comments load, that is why there is a lot of attempts to join/leave in the console on hard refresh.

- Hard refreshing conversations https://d.pr/v/M7TAS2
